python/pyiceberg/io/pyarrow.py

```diff
@@ -476,6 +483,202 @@ def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
+        """Override this method to perform an action immediately before visiting a field."""
+
+    def after_field(self, field: pa.Field) -> None:
+        """Override this method to perform an action immediately after visiting a field."""
+
+    def before_list_element(self, element: pa.Field) -> None:
+        """Override this method to perform an action immediately before visiting a list element."""
+
+    def after_list_element(self, element: pa.Field) -> None:
+        """Override this method to perform an action immediately after visiting a list element."""
+
+    def before_map_key(self, key: pa.Field) -> None:
+        """Override this method to perform an action immediately before visiting a map key."""
+
+    def after_map_key(self, key: pa.Field) -> None:
+        """Override this method to perform an action immediately after visiting a map key."""
+
+    def before_map_value(self, value: pa.Field) -> None:
+        """Override this method to perform an action immediately before visiting a map value."""
+
+    def after_map_value(self, value: pa.Field) -> None:
+        """Override this method to perform an action immediately after visiting a map value."""
+
+    @abstractmethod
+    def schema(self, schema: pa.Schema, field_results: List[T]) -> Schema:
+        """visit a schema"""
+
+    @abstractmethod
+    def struct(self, struct: pa.StructType, field_results: List[T]) -> T:
+        """visit a struct"""
+
+    @abstractmethod
+    def list(self, list_type: pa.ListType, element_result: T) -> T:
+        """visit a list"""
+
+    @abstractmethod
+    def map(self, map_type: pa.MapType, key_result: T, value_result: T) -> T:
+        """visit a map"""
+
+    @abstractmethod
+    def primitive(self, primitive: pa.DataType) -> T:
+        """visit a primitive type"""
+
+
+def _get_field_id(field: pa.Field) -> int:
+    if field.metadata is not None:
+        field_metadata = {k.decode(): v.decode() for k, v in field.metadata.items()}
+        if field_id := field_metadata.get("PARQUET:field_id"):
+            return int(field_id)
+    raise ValueError(f"Field {field.name} does not have a field_id")
+
+
+class _ConvertToIceberg(PyarrowSchemaVisitor[IcebergType], ABC):
+    def schema(self, schema: pa.Schema, field_results: List[IcebergType]) -> Schema:
+        fields = []
+        for i in range(len(schema.names)):
+            field = schema.field(i)
+            field_id = _get_field_id(field)
+            field_type = field_results[i]
+            if field_type is not None:
+                fields.append(NestedField(field_id, field.name, field_type, required=not field.nullable))
+        return Schema(*fields)
+
+    def struct(self, struct: pa.StructType, field_results: List[IcebergType]) -> IcebergType:
+        fields = []
+        for i in range(struct.num_fields):
+            field = struct[i]
+            field_id = _get_field_id(field)
+            # may need to check doc strings
```

Comment:
I re-consider this and think we do not need to fetch the optinal doc from the metadata as the inferred schema is only used to determine the correspondance between field name and field id.

Reply:
I don't think that this should assume how the converted schema will be used. It could easily be used to create a table, in which case you'd want to have docstrings.

Reply:
Thank you for your explanation. I added relevant code to handle the optional doc if exists.

I did some tests and found that optional doc seemed not to be stored in parquet field metadata:
The table is created by :
```python
spark.sql("CREATE TABLE demo.iceberg_ref.schema_doc2 (id bigint COMMENT 'unique id', data string) USING iceberg")
```
Fetch parquet schema:
```python
doc_test_path = "/Users/jonasjiang/.CMVolumes/gluetestjonas/warehouse/iceberg_ref.db/schema_doc2/data/00000-0-49fb8801-9660-47ee-a8c7-b2af7fb703b4-00001.parquet"
with fs.open_input_file(doc_test_path) as f:
    parquet_schema = pq.read_schema(f)
    print("=======parquet schema=======")
    print(parquet_schema)
    print("=======iceberg schema=======")
    print(Schema.parse_raw(parquet_schema.metadata[b'iceberg.schema'].decode()))
```
Result:
```python
=======parquet schema=======
id: int64 not null
  -- field metadata --
  PARQUET:field_id: '1'
data: string not null
  -- field metadata --
  PARQUET:field_id: '2'
-- schema metadata --
iceberg.schema: '{"type":"struct","schema-id":0,"fields":[{"id":1,"name":' + 110
=======iceberg schema=======
table {
  1: id: required long (unique id)
  2: data: required string
}
```